src/mame/machine/dmac3.h

```diff
@@ -0,0 +1,193 @@
+// license:BSD-3-Clause
+// copyright-holders:Brice Onken,Tsubai Masanari
+// thanks-to:Patrick Mackinlay
+
+/*
+ * Sony CXD8403Q DMAC3 DMA controller
+ *
+ * Register definitions were derived from the NetBSD source code, copyright (c) 2000 Tsubai Masanari.
+ *
+ * References:
+ *  - https://github.com/NetBSD/src/blob/trunk/sys/arch/newsmips/apbus/dmac3reg.h
+ *  - https://github.com/NetBSD/src/blob/trunk/sys/arch/newsmips/apbus/dmac3var.h
+ *  - https://github.com/NetBSD/src/blob/trunk/sys/arch/newsmips/apbus/dmac3.c
+ *  - https://github.com/NetBSD/src/blob/trunk/sys/arch/newsmips/apbus/spifi.c
+ */
+
+#ifndef MAME_MACHINE_DMAC3_H
+#define MAME_MACHINE_DMAC3_H
+
+#pragma once
+
+class dmac3_device : public device_t
+{
+public:
+	dmac3_device(machine_config const &mconfig, char const *tag, device_t *owner, u32 clock);
+
+	// DMAC3 has two controllers on-chip
+	// The 5000X uses controllers 0 and 1 for SPIFI/SCSI bus 0 and 1 respectively
+	enum dmac3_controller
+	{
+		CTRL0 = 0,
+		CTRL1 = 1,
+	};
+
+	// Address map setup
+	template <typename... T>
+	void set_apbus_address_translator(T &&...args) { m_apbus_virt_to_phys_callback.set(std::forward<T>(args)...); }
+	template <dmac3_controller controller>
+	void map(address_map &map)
+	{
+		map(0x0, 0x3).rw(FUNC(dmac3_device::csr_r<controller>), FUNC(dmac3_device::csr_w<controller>));
+		map(0x4, 0x7).rw(FUNC(dmac3_device::intr_r<controller>), FUNC(dmac3_device::intr_w<controller>));
+		map(0x8, 0xb).rw(FUNC(dmac3_device::length_r<controller>), FUNC(dmac3_device::length_w<controller>));
+		map(0xc, 0xf).rw(FUNC(dmac3_device::address_r<controller>), FUNC(dmac3_device::address_w<controller>));
+		map(0x10, 0x13).rw(FUNC(dmac3_device::conf_r<controller>), FUNC(dmac3_device::conf_w<controller>));
+	}
+
+	// Signal routing
+	template <typename T>
+	void set_bus(T &&tag, int spacenum) { m_bus.set_tag(std::forward<T>(tag), spacenum); }
+	template <dmac3_controller Controller>
+	auto dma_r_cb() { return m_dma_r[Controller].bind(); }
+	template <dmac3_controller Controller>
+	auto dma_w_cb() { return m_dma_w[Controller].bind(); }
+	auto irq_out() { return m_irq_handler.bind(); }
+
+	template <dmac3_controller Controller>
+	void irq_w(int state)
+	{
+		if (state)
+		{
+			m_controllers[Controller].intr |= INTR_INT;
+		}
+		else
+		{
+			m_controllers[Controller].intr &= ~INTR_INT;
+		}
+		m_irq_check->adjust(attotime::zero);
+	}
+
+	template <dmac3_controller Controller>
+	void drq_w(int state)
+	{
+		m_controllers[Controller].drq = (state != 0);
+		m_dma_check->adjust(attotime::zero);
+	}
+
+protected:
```

Comment:
Please clean up the order of members within the `protected` section – types (including enums), data members, and and member functions should each be grouped, with types placed before other members.